contracts/javascore/aggregator/src/main/java/relay/aggregator/RelayAggregator.java

```diff
@@ -0,0 +1,305 @@
+                    Integer pos = existingRelayers.get(relayerToBeRemoved);
+                    relayers.set(pos, top);
+                }
+                existingRelayers.remove(relayerToBeRemoved);
+            }
+        }
+    }
+
+    @External
+    public void registerPacket(
+            String srcNetwork,
+            String contractAddress,
+            BigInteger srcSn,
+            BigInteger srcHeight,
+            String dstNetwork,
+            byte[] data) {
+
+        adminOnly();
+
+        Packet pkt = new Packet(srcNetwork, contractAddress, srcSn, srcHeight, dstNetwork, data);
+        String id = pkt.getId();
+
+        Context.require(packets.get(id) == null, "Packet already exists");
+
+        packets.set(id, pkt);
+
+        PacketRegistered(
+                pkt.getSrcNetwork(),
+                pkt.getContractAddress(),
+                pkt.getSrcSn(),
+                pkt.getSrcHeight(),
+                pkt.getDstNetwork(),
+                pkt.getData());
+    }
+
+    @External
+    public void acknowledgePacket(
+            String srcNetwork,
+            String contractAddress,
+            BigInteger srcSn,
+            byte[] signature) {
+
+        relayersOnly();
+
+        String pktID = Packet.createId(srcNetwork, contractAddress, srcSn);
+        Packet pkt = packets.get(pktID);
+        Context.require(pkt != null, "Packet not registered");
+
+        byte[] existingSign = signatures.at(pktID).get(Context.getCaller());
+        Context.require(existingSign == null, "Signature already exists");
+
+        setSignature(pktID, Context.getCaller(), signature);
+
+        if (signatureThresholdReached(pktID)) {
+            byte[][] sigs = getSignatures(srcNetwork, contractAddress, srcSn);
+            byte[] encodedSigs = serializeSignatures(sigs);
+            PacketAcknowledged(
+                    pkt.getSrcNetwork(),
+                    pkt.getContractAddress(),
+                    pkt.getSrcSn(),
+                    pkt.getSrcHeight(),
+                    pkt.getDstNetwork(),
+                    pkt.getData(),
+                    encodedSigs);
+            removePacket(pktID);
+        }
+    }
+
+    private byte[][] getSignatures(String srcNetwork, String contractAddress, BigInteger srcSn) {
+        String pktID = Packet.createId(srcNetwork, contractAddress, srcSn);
+        DictDB<Address, byte[]> signDict = signatures.at(pktID);
+        ArrayList<byte[]> signatureList = new ArrayList<byte[]>();
+
+        for (int i = 0; i < relayers.size(); i++) {
+            Address relayer = relayers.get(i);
+            byte[] sign = signDict.get(relayer);
+            if (sign != null) {
+                signatureList.add(sign);
+            }
+        }
+
+        byte[][] sigs = new byte[signatureList.size()][];
+        for (int i = 0; i < signatureList.size(); i++) {
+            sigs[i] = signatureList.get(i);
+        }
+        return sigs;
+    }
```

Comment:
this can be optimized using branch db where  packetId and relayer address becomes key and signature becomes value.  Then you can get all signatures for pkt id or signature from one relayer for that pktid as well. look fo impl of branch db on some other contracts.

Reply:
Yeah, we already have this implemented here:
```java
private final BranchDB<String, DictDB<Address, byte[]>> signatures = Context.newBranchDB("signatures",
            byte[].class);```